pkg/distributor/ha_tracker_test.go

```diff
@@ -70,6 +70,53 @@ func checkReplicaTimestamp(t *testing.T, duration time.Duration, c *haTracker, u
 	})
 }
 
+func TestHATrackerCacheSyncOnStart(t *testing.T) {
```

Comment:
just double-checking, but this test passes on `main` too. You know that, right? My understanding is that it's just asserting that the tracker works on distributor (re)starts without losing state, not the code you just added.

Can you think of a way to test the new code too? Maybe somehow making sure that given a running HA tracker and querying an already cached replica, the HA tracker does not query the KV store again upon checkReplica?

Reply:
Yes, the test passes on the main branch as well. You are  right.

I'm struggling to find a way to isolate and test just this function 🤔. When a new ha_tracker is created, the sync* function gets called automatically, initializing the cache. Without this function, checkReplica would still update the cache.

This is why I mentioned that the addition doesn’t add much value for initializing the cache. It justs cause initialization to happen earlier. 

I'll keep looking for alternatives, but any suggestions are very welcome!

Reply:
some ideas below, _but i also don't want to block this PR on it. You can merge this and then follow-up with more tests or continue the discussion_

you can implement a decorator `kv.Client` and pass that the KV store instead of the consul one here: 

https://github.com/grafana/mimir/blob/c7e1819b5b5c622ad50317a06e091f1c20116f09/pkg/distributor/ha_tracker_test.go#L88


in your implementation you can delegate to a consul client to do the heavy-lifting of a KV-store, but note when calls are being made. For example, you can assert that after startup, no calls to `Get` or `CAS` are made (within the update timeout for example).

That way you assert that the warmup during startup was effective and we didn't go to the KV store because of a cold cache.

Reply:
Related PR: https://github.com/grafana/mimir/pull/9905